packages/jaeger-ui/src/model/transform-trace-data.tsx

```diff
@@ -107,17 +107,21 @@ export default function transformTraceData(data: TraceData & { spans: SpanWithPr
     span.relativeStartTime = span.startTime - traceStartTime;
     span.depth = depth - 1;
     span.hasChildren = node.children.length > 0;
-    const tagsInfo = deduplicateTags(span.tags);
-    span.tags = tagsInfo.tags;
     span.warnings = span.warnings || [];
-    span.warnings = span.warnings.concat(tagsInfo.warnings);
-    span.references.forEach(ref => {
-      const refSpan = spanMap.get(ref.spanID) as Span;
-      if (refSpan) {
-        // eslint-disable-next-line no-param-reassign
-        ref.span = refSpan;
-      }
-    });
+    if (Array.isArray(span.tags)) {
+      const tagsInfo = deduplicateTags(span.tags);
+      span.tags = tagsInfo.tags;
+      span.warnings = span.warnings.concat(tagsInfo.warnings);
+    }
+    if (Array.isArray(span.references)) {
```

Comment:
The [type def for `SpanData`](https://github.com/jaegertracing/jaeger-ui/blob/master/packages/jaeger-ui/src/types/trace.tsx#L55-L56) has `references` and `tags` as required arrays, and other parts of the code base assume they exist. I think it would be best to sanitize the trace data rather than necessitate this check whenever these values are accessed.

Adding `span.references = span.references || [];` above `if (!span.references || !span.references.length) {` (and then changing that line to `if (!span.references.length) {`) as well as adding `span.tags = span.tags || [];` above `const tagsInfo = deduplicateTags(span.tags);` would follow the pattern started by `span.warnings` and remove the need for `Array.isArray` checks.

The types for `SpanData` and `Span` are currently somewhat misleading, both before this PR and now a bit more.
I would change the types in `src/types/trace.tsx` to [better capture when certain properties are optional](https://www.typescriptlang.org/play/#src=%2F%2F%20Stand%20in%20types%0D%0Atype%20KeyValuePair%20%3D%20string%3B%0D%0Atype%20Log%20%3D%20string%3B%0D%0Atype%20Process%20%3D%20string%3B%0D%0Atype%20SpanReference%20%3D%20string%3B%0D%0A%0D%0Atype%20SpanData%20%3D%20%7B%0D%0A%20%20%20%20spanID%3A%20string%3B%0D%0A%20%20%20%20traceID%3A%20string%3B%0D%0A%20%20%20%20processID%3A%20string%3B%0D%0A%20%20%20%20operationName%3A%20string%3B%0D%0A%20%20%20%20startTime%3A%20number%3B%0D%0A%20%20%20%20duration%3A%20number%3B%0D%0A%20%20%20%20logs%3A%20Array%3CLog%3E%3B%0D%0A%20%20%20%20process%3A%20Process%3B%20%2F%2F%20Note%20this%20was%20moved%20to%20SpanData%20from%20Span%0D%0A%20%20%20%20tags%3F%3A%20Array%3CKeyValuePair%3E%3B%0D%0A%20%20%20%20references%3F%3A%20Array%3CSpanReference%3E%3B%0D%0A%20%20%20%20warnings%3F%3A%20Array%3Cstring%3E%3B%0D%0A%7D%3B%0D%0A%0D%0Atype%20Span%20%3D%20SpanData%20%26%20%7B%0D%0A%20%20%20%20depth%3A%20number%3B%0D%0A%20%20%20%20hasChildren%3A%20boolean%3B%0D%0A%20%20%20%20relativeStartTime%3A%20number%3B%0D%0A%0D%0A%20%20%20%20%2F%2F%20Exclude%3CSpanData%5B'optionalProperty'%5D%2C%20undefined%3E%20and%20NonNullable%3CSpanData%5B'optionalProperty'%5D%3E%20both%20make%20optional%20property%20of%20SpanData%20required%0D%0A%20%20%20%20%2F%2F%20I%20just%20learned%20of%20these%20conditional%20types%2C%20so%20I'm%20not%20sure%20if%20one%20is%20preferred%20over%20the%20other%2C%20but%20I'd%20lean%20towards%20NonNullable%0D%0A%20%20%20%20%2F%2F%20More%20info%3A%20https%3A%2F%2Fgithub.com%2Fmicrosoft%2FTypeScript%2Fpull%2F21847%20%0D%0A%20%20%20%20tags%3A%20Exclude%3CSpanData%5B'tags'%5D%2C%20undefined%3E%3B%0D%0A%20%20%20%20references%3A%20NonNullable%3CSpanData%5B'references'%5D%3E%3B%0D%0A%20%20%20%20warnings%3A%20NonNullable%3CSpanData%5B'warnings'%5D%3E%3B%0D%0A%7D%3B%0D%0A%0D%0Aconst%20spanDataNotSpan%3A%20Span%20%3D%20%7B%0D%0A%20%20%20%20spanID%3A%20'sID'%2C%0D%0A%20%20%20%20traceID%3A%20'tID'%2C%0D%0A%20%20%20%20processID%3A%20'pID'%2C%0D%0A%20%20%20%20operationName%3A%20'opName'%2C%0D%0A%20%20%20%20startTime%3A%2015000%2C%0D%0A%20%20%20%20duration%3A%20300%2C%0D%0A%20%20%20%20logs%3A%20%5B%5D%2C%0D%0A%20%20%20%20process%3A%20'would%20be%20a%20process'%2C%0D%0A%20%20%20%20depth%3A%202%2C%0D%0A%20%20%20%20hasChildren%3A%20false%2C%0D%0A%20%20%20%20relativeStartTime%3A%20200%2C%0D%0A%0D%0A%20%20%20%20%2F%2F%20type%20errors%20due%20to%20being%20undefined%2C%20requires%20strictNullChecks%20to%20be%20enabled%20under%20options%2C%20which%20JaegerUI%20uses%0D%0A%20%20%20%20tags%3A%20undefined%2C%0D%0A%20%20%20%20references%3A%20undefined%2C%0D%0A%20%20%20%20warnings%3A%20undefined%2C%0D%0A%7D). Then `SpanData` could be used where `SpanWithProcess` is currently used.

Reply:
Alternatively, the `else` could do this.

```js
if (Array.isArray(span.tags)) {
  const tagsInfo = deduplicateTags(span.tags);
  span.tags = tagsInfo.tags;
  span.warnings = span.warnings.concat(tagsInfo.warnings);
} else {
  span.tags = [];
}
```

@rubenvp8510 thanks for the PR. Any thoughts about @everett980 remarks?

Reply:
Hi, sorry I was busy and abandoned this PR a little bit.

I think it makes sense to sanitize the values on this part of the code (and do some asumptions on the rest) , Also the changes for the types  could help in the future to prevent more issues with null/undefined values.

I'll update this PR with the proposed changes.